service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,160 @@
+
+        if (!giftCertificateValidator.isGiftCertificateCreationFormValid(giftCertificate)) {
+            throw new InvalidFieldException();
+        }
+
+        giftCertificate.setCreateDate(LocalDateTime.now(ZoneId.of(TIME_ZONE)));
+        giftCertificate.setLastUpdateDate(LocalDateTime.now(ZoneId.of(TIME_ZONE)));
+
+        checkForTags(giftCertificate.getTags());
+
+        long lastAddedId = giftCertificateDao.create(giftCertificate);
+        giftCertificateTagDao.createConnections(lastAddedId, giftCertificate.getTags());
+    }
+
+    @Override
+    @Transactional
+    public void updateGiftCertificate(long id, GiftCertificateDto giftCertificateDto) throws InvalidFieldException,
+            ResourceNotFoundException {
+
+        if (giftCertificateDao.findById(id).isEmpty()) {
+            throw new ResourceNotFoundException(id);
+        }
+
+        GiftCertificate giftCertificate = giftCertificateDtoWrapper.unwrap(giftCertificateDto);
+
+        if (!giftCertificateValidator.isGiftCertificateCreationFormValid(giftCertificate)) {
+            throw new InvalidFieldException();
+        }
+
+        giftCertificate.setLastUpdateDate(LocalDateTime.now(ZoneId.of(TIME_ZONE)));
+
+        checkForTags(giftCertificate.getTags());
+
+        if (giftCertificate.getTags().isEmpty()){
+            giftCertificateTagDao.deleteConnectionBetweenGiftCertificateAndTagsByGiftCertificateId(id);
+        }
+        else {
+            connectCertificatesAndTags(id, giftCertificate);
+        }
+
+        giftCertificateDao.update(id, giftCertificate);
+    }
+
+
+    @Override
+    @Transactional
+    public void deleteGiftCertificate(long id) throws ResourceNotFoundException {
+        if (giftCertificateDao.findById(id).isEmpty()) {
+            throw new ResourceNotFoundException(id);
+        }
+        giftCertificateTagDao.deleteConnectionBetweenGiftCertificateAndTagsByGiftCertificateId(id);
+        giftCertificateDao.delete(id);
+    }
+
+    public Set <Tag> getTagsByGiftCertificateId(long id) throws ResourceNotFoundException {
+        List <Tag> listOfTags = tagDao.getAllTagsConnectedWithCertificateId(id);
+        Set<Tag> setOfTags = new HashSet<>(listOfTags);
+        return setOfTags;
+    }
+
+    private void checkForTags(Set <Tag> newTags) {
+        List<String> newTagNames = new ArrayList<>();
+
+        newTags.stream().forEach(tag -> newTagNames.add(tag.getName()));
+        newTagNames.stream().filter(tagName -> tagDao.findByName(tagName).isEmpty())
+                .forEach(tagName -> tagDao.create(new Tag(tagName)));
+    }
+
+    private void connectCertificatesAndTags(long id, GiftCertificate giftCertificate) {
+        List<Long> tagIdsBeforeUpdate = giftCertificateTagDao.getAllTagIdConnectedWithCertificateId(id);
+        List<Long> tagIdsAfterUpdate = tagDao.getAllTagsIdConnectedWithGiftCertificate(giftCertificate.getTags());
+
+
+        for (Long tagId : tagIdsAfterUpdate){
```

Comment:
"For" can be replaced by java stream

Reply:
change for tagIdsAfterUpdate.stream().filter(tagId->!tagIdsBeforeUpdate.contains(tagId)).
                forEach(tagId -> giftCertificateTagDao.addTagId(id,tagId));